Add filter inference, normalization and push-down logic to v4

rytaft opened the pull request:
This PR makes a few small updates to the transformation logic
in the `v4` version of opttoy:
1. Normalize comparison expressions so the variable is always on
   the left.  E.g., `5 > x` -> `x < 5`. This was already the case for
   equals and not-equals (= and !=), but this commit adds the
   inequalities <, <=, >, and >=.
2. Infer filters on equivalent columns. For example, if two columns
   x and y are equal, and there is already a filter on `x` (e.g. `x < 5`),
   then an additional filter is added for `y` (`y < 5`).
3. Fix the push-down logic so that multiple filters on the same table
   get pushed down to the same subquery. Previously, a new subquery was
   created for each pushed-down filter.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/petermattis/opttoy/54)
<!-- Reviewable:end -->

andy-kimball commented on v4/opt/factory.go:
```diff
@@ -245,6 +264,84 @@ func (f *Factory) appendColumnProjections(projections, group GroupID) GroupID {
 
+// substitute recursively substitutes oldCol with newCol in the given filter
+// expression. For example, if oldCol=x and newCol=y, x < 5 will become y < 5.
+func (f *Factory) substitute(filter GroupID, oldCol, newCol ColumnIndex) GroupID {
```
This looks correct to me. Maybe add a TODO to improve performance before we add into production code. In particular, it's important that if we don't find oldCol in a particular subtree, that we don't reconstruct the node. Ideally, we'd code generate some kind of "replace visitor" that would do the heavy lifting of rebuilding a tree bottom-up without creating unnecessary extra nodes.

andy-kimball commented on v4/opt/factory.go:
```diff
@@ -88,7 +107,7 @@ func (f *Factory) concatFilterConditions(filterLeft, filterRight GroupID) GroupI
 }
 
-func (f *Factory) flattenFilterCondition(filter GroupID) GroupID {
+func (f *Factory) flattenFilterCondition(input ListID, filter GroupID) GroupID {
```
Adding the check here might miss some cases. The trouble is that `flattenFilterCondition` is called once in order to turn a possibly nested select/join predicate into a flattened list. But what happens if other patterns add items to that list later on (e.g. pushdown patterns)? It wouldn't go through this code path, and so we wouldn't infer equivalent filters for the newly added expressions.

I think this is fine for opttoy, but maybe add a comment about the issue, to be solved later? We might decide to add separate patterns for inferring equivalent filters (for both select and join cases). I'm not quite sure how we can efficiently check whether we've already pushed down filters, because if we're not careful, we could repeatedly try to push down the same filters. Your solution avoids this.

andy-kimball commented on v4/opt/norm/push_down.opt:
```diff
@@ -13,9 +13,24 @@
 
+[PushDownSelectJoinSelectLeft, Normalize]
+(Select
+    $input:(Join (Select $inputInner:* (Filters [ ... $filterInner:* ... ])) $right:* $on:*)
```
The `[ ... $filterInnder:* ... ]` pattern will match the first item in the list (because you used * to match). But what if there is more one item in the list? Those other items would get discarded by the replace pattern.

I think I see why you added this pattern - my default pushdown pattern would wrap a `Select` with a `Select`, and that pattern avoids doing that. However, I think it'd be better to create a pattern similar to this:
```
[MergeSelectSelect, Normalize]
(Select
  (Select $input $innerFilter:*)
  $outerFilter:*
)
=>
(Select
  $input
  (ConcatFilterConditions $outerFilter $innerFilter)
)
```

Because of the way the factory works, this is going to be fairly efficient, as we don't actually construct the outer `Select`. Also, this pattern will match in many other useful cases besides Pushdown.

RaduBerinde commented:
Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[v4/opt/factory.go, line 336 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/54#-L2vf-vPz2P8fras6qE7:-L2vf-vPz2P8fras6qE8:b-iv40mw) ([raw file](https://github.com/petermattis/opttoy/blob/5d8fd4a1e8900fac595a1e600b9fe025579730fd/v4/opt/factory.go#L336)):*
> ```Go
> 		for _, equivCols := range equivColSets {
> 			filterCols := f.mem.lookupGroup(filter).logical.UnboundCols
> 			if filterCols.SubsetOf(equivCols) {
> ```

I don't understand what's going on here.. All columns in the filter have to be equivalent in order to do a substitution? What about e.g. something like `a.y + b.y = 1`? We would want to replace this with `a.y + a.x = 1` for the left side.

The way I was thinking of doing this (and kind of what we do in cockroach already with `splitFilter`) is to provide a set of "acceptable" columns along with the information about column equivalencies (as a `util.UnionFind`); with this information, we would try to find an "acceptable" equivalent column for each "unacceptable" column in the filter. This would be used for pushdown through joins (once with left-side columns as "acceptable", once with right-side).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/54)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the comments!  Please take another look.

---

Review status: 0 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[v4/opt/factory.go, line 110 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/54#-L2vJ59vumnEO-VgXOwO-r1-110:-L30H8mLNUqlwkdJXMaH:b-896fix) ([raw file](https://github.com/petermattis/opttoy/blob/5d8fd4a1e8900fac595a1e600b9fe025579730fd/v4/opt/factory.go#L110)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Adding the check here might miss some cases. The trouble is that `flattenFilterCondition` is called once in order to turn a possibly nested select/join predicate into a flattened list. But what happens if other patterns add items to that list later on (e.g. pushdown patterns)? It wouldn't go through this code path, and so we wouldn't infer equivalent filters for the newly added expressions.

I think this is fine for opttoy, but maybe add a comment about the issue, to be solved later? We might decide to add separate patterns for inferring equivalent filters (for both select and join cases). I'm not quite sure how we can efficiently check whether we've already pushed down filters, because if we're not careful, we could repeatedly try to push down the same filters. Your solution avoids this.
</blockquote></details>

Done.

---

*[v4/opt/factory.go, line 269 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/54#-L2vJ59vumnEO-VgXOwO-r1-269:-L30HAP-NqqWcEnWkfLC:b-5nwkbv) ([raw file](https://github.com/petermattis/opttoy/blob/5d8fd4a1e8900fac595a1e600b9fe025579730fd/v4/opt/factory.go#L269)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This looks correct to me. Maybe add a TODO to improve performance before we add into production code. In particular, it's important that if we don't find oldCol in a particular subtree, that we don't reconstruct the node. Ideally, we'd code generate some kind of "replace visitor" that would do the heavy lifting of rebuilding a tree bottom-up without creating unnecessary extra nodes.
</blockquote></details>

Done.

I've added some logic to prevent constructing extra nodes since it was fairly straightforward. Are there additional performance issues you see that would be improved by generated code?

---

*[v4/opt/factory.go, line 336 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/54#-L2vf-vPz2P8fras6qE7:-L30HiMGkJisanPaA43T:bdiks6k) ([raw file](https://github.com/petermattis/opttoy/blob/5d8fd4a1e8900fac595a1e600b9fe025579730fd/v4/opt/factory.go#L336)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't understand what's going on here.. All columns in the filter have to be equivalent in order to do a substitution? What about e.g. something like `a.y + b.y = 1`? We would want to replace this with `a.y + a.x = 1` for the left side.

The way I was thinking of doing this (and kind of what we do in cockroach already with `splitFilter`) is to provide a set of "acceptable" columns along with the information about column equivalencies (as a `util.UnionFind`); with this information, we would try to find an "acceptable" equivalent column for each "unacceptable" column in the filter. This would be used for pushdown through joins (once with left-side columns as "acceptable", once with right-side).
</blockquote></details>

Good point! I've fixed this so it handles the case you mentioned (added a test case as well).  As we discussed over Slack, this won't handle all cases and may create some unnecessary filters.  But it's probably sufficient for now...

---

*[v4/opt/norm/push_down.opt, line 18 at r1](https://reviewable.io:443/reviews/petermattis/opttoy/54#-L2vJ59vumnEO-VgXOwS-r1-18:-L30IK-vMKpnVijf-opI:bdef8bz) ([raw file](https://github.com/petermattis/opttoy/blob/5d8fd4a1e8900fac595a1e600b9fe025579730fd/v4/opt/norm/push_down.opt#L18)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

The `[ ... $filterInnder:* ... ]` pattern will match the first item in the list (because you used * to match). But what if there is more one item in the list? Those other items would get discarded by the replace pattern.

I think I see why you added this pattern - my default pushdown pattern would wrap a `Select` with a `Select`, and that pattern avoids doing that. However, I think it'd be better to create a pattern similar to this:
```
[MergeSelectSelect, Normalize]
(Select
  (Select $input $innerFilter:*)
  $outerFilter:*
)
=>
(Select
  $input
  (ConcatFilterConditions $outerFilter $innerFilter)
)
```

Because of the way the factory works, this is going to be fairly efficient, as we don't actually construct the outer `Select`. Also, this pattern will match in many other useful cases besides Pushdown.
</blockquote></details>

Good idea!  I've added `MergeSelectSelect` to `norm.opt` since that seemed like a good place for it.  I basically used exactly what you pasted above -- just made a few minor changes to fix errors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/54#-:-L30H1VmuZjzzeWPsFHI:b-nmdah2)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Reviewed 3 of 11 files at r1, 8 of 8 files at r2.
Review status: 3 of 10 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/54)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 10 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/petermattis/opttoy/54#-:-L30S5_W2pH6hNejeq6J:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM
